src/support/bitset.rs

```diff
@@ -0,0 +1,161 @@
+use core::fmt;
+use std::{fmt::{Display, Formatter}, mem::size_of};
+
+type Word = u32;
+pub struct BitSet {
+    count_ones: usize,
+    buckets: Vec<Word>
+}
+
+impl BitSet {
+    pub fn new() -> Self {
+        Self {
+            count_ones: 0,
+            buckets: Vec::new(),
+        }
+    }
+
+    fn bits_in_bucket() -> usize {
+        return size_of::<Word>() * 8;
+    }
+
+    pub fn len(&self) -> usize {
+        return self.buckets.len() * Self::bits_in_bucket();
+    }
+
+    pub fn set(&mut self,  val: bool, index : usize) {
+        let num_bucket = index / Self::bits_in_bucket();
+        if num_bucket + 1 > self.buckets.len() {
+            self.buckets.resize(num_bucket + 1, 0);
+        }
+        let bit_pos = index % (Self::bits_in_bucket());
+        let flag = 1 << bit_pos;
+        if val {
+            self.buckets[num_bucket] |= flag;
+            self.count_ones += 1;
+        } else {
+            self.buckets[num_bucket] &= !flag;
+            self.count_ones -= 1;
+        }
+    }
+
+    pub fn get(&self, index : usize) -> bool {
+        if index >= self.buckets.len() * 32 {
+            return false;
+        }
+        let num_bucket = index / Self::bits_in_bucket();
+        let bit_pos = index % Self::bits_in_bucket();
+        return self.buckets[num_bucket] & (1 << bit_pos) != 0;
+    }
+
+    pub fn count_ones(&self) -> usize {
+        return self.count_ones;
+    }
+
+    pub fn count_zeros(&self) -> usize {
+        return self.len() - self.count_ones;
+    }
+
+    pub fn shirnk_zero(&mut self) {
+        let mut i = self.buckets.len() - 1;
+        while self.buckets[i] == 0 {
+            self.buckets.pop();
+            if i == 0 {
+                break;
+            }
+            i -= 1;
+        }
+    }
+}
+
```

Comment:
lets also make fn iter_ones(&self) -> Iter<usize> that iterates over indices of ones effectively, using popcnt